velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -17,41 +17,151 @@
       insertTableHandle_(std::move(insertTableHandle)),
       connectorQueryCtx_(connectorQueryCtx),
-      writeProtocol_(std::move(writeProtocol)) {
-  VELOX_CHECK_NOT_NULL(
-      writeProtocol_, "Write protocol could not be nullptr for HiveDataSink.");
-}
-
-std::shared_ptr<ConnectorCommitInfo> HiveDataSink::getConnectorCommitInfo()
-    const {
-  return std::make_shared<HiveConnectorCommitInfo>(writerParameters_);
-}
+      commitStrategy_(commitStrategy) {}
 
 void HiveDataSink::appendData(VectorPtr input) {
   // For the time being the hive data sink supports one file
   // To extend it we can create a new writer for every
   // partition
   if (writers_.empty()) {
-    writers_.emplace_back(createWriter());
+    std::shared_ptr<const HiveWriterParameters> hiveWriterParameters =
+        getWriterParameters(std::nullopt);
+    writers_.emplace_back(createWriter(hiveWriterParameters));
+    writerInfo_.emplace_back(
+        std::make_shared<HiveWriterInfo>(*hiveWriterParameters));
   }
   writers_[0]->write(input);
+  writerInfo_[0]->numWrittenRows += input->size();
+}
+
+RowVectorPtr HiveDataSink::finish() const {
+  memory::MemoryPool* pool = connectorQueryCtx_->memoryPool();
+
+  if (outputType_->size() <= 1) {
+    return std::make_shared<RowVector>(
+        pool,
+        outputType_,
+        nullptr,
+        1,
+        std::vector<VectorPtr>{BaseVector::createConstant(
+            (int64_t)computeTotalWrittenRows(writerInfo_), 1, pool)});
+  }
+
+  std::vector<VectorPtr> columns = {};
+  vector_size_t numOutputRows = writers_.size() + 1;
+
+  // Set rows column.
+  FlatVectorPtr<int64_t> writtenRowsVector =
+      BaseVector::create<FlatVector<int64_t>>(BIGINT(), numOutputRows, pool);
+  writtenRowsVector->set(0, (int64_t)computeTotalWrittenRows(writerInfo_));
+  for (int idx = 1; idx < numOutputRows; ++idx) {
+    writtenRowsVector->setNull(idx, true);
+  }
+  columns.emplace_back(writtenRowsVector);
+
+  // Set fragments column.
+  FlatVectorPtr<StringView> fragmentsVector =
+      BaseVector::create<FlatVector<StringView>>(
+          VARBINARY(), numOutputRows, pool);
+  fragmentsVector->setNull(0, true);
+  for (int i = 1; i < numOutputRows; i++) {
+    fragmentsVector->set(
+        i, StringView(makePartitionUpdate(writerInfo_[i - 1])));
+  }
+  columns.emplace_back(fragmentsVector);
+
+  // Set commitcontext column.
+  // clang-format off
+     auto commitContextJson = folly::toJson(
+       folly::dynamic::object
+           ("lifespan", "TaskWide")
+           ("taskId", connectorQueryCtx_->taskId())
+           ("pageSinkCommitStrategy", commitStrategyToString(commitStrategy_))
+           ("lastPage", true));
+  // clang-format on
+  VectorPtr commitContextVector = BaseVector::createConstant(
+      variant::binary(commitContextJson), numOutputRows, pool);
+  columns.emplace_back(commitContextVector);
+
+  return std::make_shared<RowVector>(
+      pool, outputType_, nullptr, numOutputRows, columns);
```

Comment:
The size of the three child vectors of the `TableWriter::getOutput()`, `numOutputRows` here, is the number of partitions plus 1. Only connector/HiveDataSink has this info, so I decide to assemble the three output vectors inside HiveDataSink instead of TableWriter.

I also need to pass Operator's `outputType_` to connector/DataSink/HiveDataSink, since `outputType_ `is used at the same place as `numOutputRows` as in line 164.

Also I move the logic to track the `numWrittenRows` from TableWriter to connector/HiveDataSink, since we actually need to track the per-partition `numWrittenRows`. numWrittenRows is also needed to populate the first vector.